### PR TITLE
Port E2E tests to `kubernetes-sigs/e2e-framework`

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Fixes https://github.com/crossplane/crossplane/issues/4013

Crossplane relies heavily on unit tests to ensure quality, but these alone are not enough to ensure Crossplane functions correctly as a whole. Thus far we have relied on an honor system of folks running manual black-box tests to increase the likelihood that their changes work in real-world conditions. These tests:

* Are not very repeatable, or portable for others to run.
* Are often only run on one environment - i.e. `kind` on the author's laptop.
* Are not re-run on future changes to catch regressions.
* Are likely to miss things like performance degradation.
* Are usually skipped for innocuous-seeming things like dependency bumps.
* Often don't consider interactions with features besides the one under test.
* Often don't cover upgrade scenarios.

We've made several attempts to add integration and E2E tests to Crossplane in the past, but none have flourished.

Under this renewed effort I propose:

* We move all E2E testing code into this repo - c/c.
* We run all E2E tests on every PR, breaking the build when tests fail.
* We update the contributing guide and PR template to encourage E2E tests.
* We attempt to make E2E tests low effort to add and expand.

This PR contains a one-pager and an implementation of the proposal.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR if necessary.~

### How has this code been tested

[contribution process]: https://git.io/fj2m9

It _is_ tests. 🙂 The tests did actually find https://github.com/crossplane/crossplane/pull/4188 though. Note that GitHub Actions invokes `make e2e`, so you can see that the e2e-framework based tests have run in this PR's checks.

### CODE REVIEW

**Comment by negz** [comment]
I'm leaving this in draft until the implementation I have here has parity with our existing E2E test suite.

If we accept this proposal I'd like to be able to cut over from our existing E2E tests when it's merged. This PR already removes `integration_tests.sh` and replaces `test/e2e`. If we merge it I'll also archive https://github.com/crossplane/test.

To have parity I need to test:

- [ ] That the Crossplane deployment starts successfully when installed.
- [ ] That it's possible upgrade Crossplane from the stable Helm chart to a master build.
- [ ] That it's possible to create a claim using a very basic Composition.
- [ ] That it's possible upgrade a Provider from one version to another.
- [ ] That it's possible to resolve a Configuration package's dependencies.

**Comment by negz** [comment]
This is now ready for review. Given that it's introducing a new way to write tests, and we'll all have to write those tests, I'm definitely happy to tweak the approach if there's anything folks are concerned with.

CI isn't quite passing because `CI / e2e-tests` is required, but this PR runs a matrix of tests like `CI / e2e-tests (install)` instead. If this PR is approved I'll tweak the set of required checks to include all of the e2e-test variants.

I have a few things I want to improve on that I don't think need to block this PR. Namely:

- Update to https://github.com/crossplane-contrib/provider-nop/pull/8
- Push all test Configuration packages to a new xpkg.upbound.io/crossplane-e2e org.
- Make it possible to run the tests against and existing cluster (i.e. skip creating a cluster and deploying Crossplane)
- Make it more obvious when you're trying to use a manifest that doesn't exist (this got me a few times)

File: test/e2e/README.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+1. A directory of manifests - i.e. test fixtures.
+1. A `features.Table` of assessment steps.
+
+Features are grouped into broad feature areas - e.g. `TestComposition` in
+`composition_test.go`. Features pertaining to Composition should be added to
+`TestComposition`.
+
+Some tests may involve updating an existing test - for example you might test a
+new kind of transform by updating the `composition/patch-and-transform`
+manifests and adding a new assessment or two to the `pandt` `features.Table`.
```
Comment (turkenh) [inline test/e2e/README.md:67]: Given we would require adding or updating e2e tests for every PR, I am curious how much the `features.Table` will grow, especially for features like Patch and Transform, where a lot of different cases need to be tested.

**Review by turkenh — APPROVED** [review]
Thanks @negz!

I really liked the declarative nature of the new testing structure and the batteries included in the proposed testing framework for testing Kubernetes controllers. I left some comments, but none of them is blocking!

File: design/one-pager-e2e-tests.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+Because it's run in a separate repository and isn't connected to PRs or releases
+it's unclear whether anyone notices if/when tests break. It's also easy to
+forget to update the tests when adding new functionality.
+
+## Goals
+
+The goal of this proposal is to:
+
+* Automatically simulate real-world use of all of Crossplane's functionality.
+* Catch the kind of bugs that unit tests miss, before they're released.
```
[the remark this answers is not in the document]
Comment (negz) [reply design/one-pager-e2e-tests.md:67]: Definitely - will update.

I do think eventually it might be tough to run the entire suite on every PR, but we can address that when we get to it. (Maybe add a `/test` trigger similar to upjet, Kube, etc).

File: test/e2e/apiextensions_test.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+				funcs.ResourcesDeletedWithin(3*time.Minute, manifests, "prerequisites/*.yaml"),
+			),
+		},
+	}
+
+	setup := funcs.ReadyToTestWithin(1*time.Minute, namespace)
+	environment.Test(t,
+		minimal.Build("Minimal").
+			WithLabel("area", "apiextensions").
+			WithLabel("size", "small").
```
[the remark this answers is not in the document]
Comment (negz) [reply test/e2e/apiextensions_test.go:128]: I played around with that, but reverted it for some reason. Let me take another pass. I'll either make them constants or add a TODO to try again in future.

File: .github/workflows/pr.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    types: [opened, edited, synchronize]
+
+jobs:
+  checklist-completed:
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: mheap/require-checklist-action@v2
+        with:
+          # The checklist must _exist_ and be filled out.
+          requireChecklist: false
```
[the remark this answers is not in the document]
Comment (negz) [reply .github/workflows/pr.yml:14]: Yes, good catch.

File: Makefile [hunk 5]
```diff
@@ -115,14 +116,23 @@ cobertura:
...
 
-# Run integration tests.
-test-integration: $(KIND) $(KUBECTL) $(HELM3)
-	@$(INFO) running integration tests using kind $(KIND_VERSION)
-	@$(ROOT_DIR)/cluster/local/integration_tests.sh || $(FAIL)
-	@$(OK) integration tests passed
+# NOTE(negz): There's already a go.test.integration target, but it's weird.
+# This relies on make build building the e2e binary.
+E2E_TEST_FLAGS ?=
+e2e-run-tests: $(KIND) $(HELM3)
```
[the remark this answers is not in the document]
Comment (negz) [reply Makefile:128]: I didn't realize that. Will look into it.

File: test/e2e/README.md (hunk 1 above)
Comment (negz) [reply test/e2e/README.md:67]: I agree. For P&T specifically I was thinking we could perhaps test many patches and transforms as one 'feature' rather than breaking them all up. I.e. Feature == P&T rather than feature == toUpper.

I've added a note that this pattern isn't set in stone though - we're still learning. I do want us to discuss and evolve consciously though.

File: cluster/local/integration_tests.sh
[the remark this answers is not in the document]
Comment (negz) [reply cluster/local/integration_tests.sh:1]: Not 100% sure I follow.

`./kind.sh` still exists, so that flow is still possible separately from the E2E tests.

WRT the tests tearing down the cluster - I do think it would be useful to be able to opt-out of that. I often find myself wanting to connect to `kind` and debug (look at logs etc) when something goes wrong. I think we can track that as a future improvement though.

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (negz) [reply Makefile:128]: I've hacked around this by creating symlinks and adding them to `$PATH` before running the test. I raised https://github.com/kubernetes-sigs/e2e-framework/issues/282 to make the binaries configurable. So far the e2e-framework folks have been quite responsive so I'm hopeful we'll get that functionality.

File: cluster/local/integration_tests.sh
[the remark this answers is not in the document]
Comment (turkenh) [reply cluster/local/integration_tests.sh:1]: Ah, ok. I was confused and thought that we were using this file there.
